### PR TITLE
[ICDS Dashboard]: Fix the Availing Services in the Child health

### PR DESCRIPTION
**Pull request by ShubhamBansal1997**
##### SUMMARY

PR in continuation with this https://github.com/dimagi/commcare-hq/pull/26827
JIRA Ticket: https://dimagi-dev.atlassian.net/browse/ICDS-1229 

Implement the new availing services in the child health monthly to calculate the registration status, which is previously calculated using the person cases.

### CODE REVIEW

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 1]
```diff
@@ -93,8 +94,12 @@ def _state_aggregation_query(self, state_id):
         )
         migration_status = "(agg_migration.is_migrated=1 AND agg_migration.migration_date < {})::integer".format(
             start_month_string)
-        seeking_services = "(person_cases.registered_status IS DISTINCT FROM 0 AND {} IS DISTINCT FROM 1)".format(
-            migration_status)
+        availing_services = (
+            "CASE WHEN NOT (agg_as.is_registered=0 AND agg_as.registration_date < {})"
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: This would be a lot nicer if it used a named parameter in the string formatting  (same with the above and below queries)

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 2]
```diff
@@ -93,8 +94,12 @@ def _state_aggregation_query(self, state_id):
...
+            "THEN 1 ELSE 0 END"
+        ).format(start_month_string)
+        seeking_services = "(person_cases.registered_status IS DISTINCT FROM 0 AND {} IS DISTINCT FROM 1 AND {} IS DISTINCT FROM 0)".format(
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: what is the distinction here between seeking services and availing services? we use those phrases interchangeably in a lot of places so its very confusing to have them be distinct here

**Review by calellowitz — CHANGES_REQUESTED** [review]

**Comment by calellowitz** [comment]
@ShubhamBansal1997 When you force push like this, it makes it hard to tell what changes are new since the prior review. It is generally good practice not to change history once review starts and only add new commits

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 3]
```diff
@@ -369,6 +373,10 @@ def _state_aggregation_query(self, state_id):
               AND agg_migration.month = %(start_date)s
               AND child_health.state_id = agg_migration.state_id
               AND child_health.supervisor_id = agg_migration.supervisor_id
+            LEFT OUTER JOIN "{agg_as_table}" agg_as ON child_health.mother_id = agg_as.person_case_id
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: nit: agg_as is not an intuitive name. this might be clearer if you spelled it out a bit more, maybe `agg_availing`

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 4]
```diff
@@ -91,10 +92,13 @@ def _state_aggregation_query(self, state_id):
         alive_in_month = "(child_health.date_death IS NULL OR child_health.date_death - {} >= 0)".format(
             start_month_string
         )
-        migration_status = "(agg_migration.is_migrated=1 AND agg_migration.migration_date < {})::integer".format(
+        migration_status = "(agg_migration.is_migrated = 1 AND agg_migration.migration_date < {})::integer".format(
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: would be good to change this to use a named parameter as well

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 5]
```diff
@@ -91,10 +92,13 @@ def _state_aggregation_query(self, state_id):
...
             start_month_string)
-        seeking_services = "(person_cases.registered_status IS DISTINCT FROM 0 AND {} IS DISTINCT FROM 1)".format(
-            migration_status)
+        registered_status = (
+            "CASE WHEN agg_as.is_registered = 0 AND agg_as.registration_date::date < {start_month_string}"
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: Can you explain in plain english what this logic is supposed to do. This logic seems quite odd to me.
This would also be much clearer as a real boolean rather than a case statement with an integer.

Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: Can registered status ever be null? This might behave oddly if so

**Review by calellowitz — COMMENTED** [review]
Nothing blocking but a few questions before approval

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py (hunk 3 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: Done!

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py (hunk 4 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: Done!

**Comment by ShubhamBansal1997** [comment]
@calellowitz Actually you are observing too many force push because these branches are made from other branches so I rebase them from the base before pushing due to which I have to do a force push every time.

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py (hunk 5 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: Done!

**Comment by calellowitz** [comment]
@ShubhamBansal1997 If you merge instead of rebase, you will not have to force push, and it will be clear that you have merged in another branch. Out of curiosity, why are they all based on other branches. As you are experiencing, that increases the complexity of managing these PRs a lot, and since these are being merged into master, do not seem necessary (beyond regular master merges if conflicts occur)

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py (hunk 5 above)
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: This was marked done, but the comment about integers vs booleans was not addressed. Although again, this hard to tell with the rebases

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 6]
```diff
@@ -92,13 +92,16 @@ def _state_aggregation_query(self, state_id):
...
             start_month_string
         )
-        migration_status = "(agg_migration.is_migrated = 1 AND agg_migration.migration_date::date < {" \
-                           "start_month_string})::integer".format(start_month_string=start_month_string)
-        registered_status = "(agg_availing.is_registered = 0 AND agg_availing.registration_date::date < {" \
-                            "start_month_string})::integer".format(start_month_string=start_month_string)
-        seeking_services = "({registered_status} IS DISTINCT FROM 1 AND " \
-                           "{migration_status} IS DISTINCT FROM 1)".format(registered_status=registered_status,
-                                                                           migration_status=migration_status)
+        migration_status = (
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py:95]: these should have parentheses (like they used to). i suspect that is at least part of the test issue you are seeing

Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py:95]: It might be clearer also to name these "migrated" and "registered" (or i guess in this case not_registered) so that it is obvious what the boolean means. "migration_status" does not tell you whether True means migrated or not migrated

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 7]
```diff
@@ -92,15 +92,11 @@ def _state_aggregation_query(self, state_id):
...
-        not_registered = (
-            "agg_availing.is_registered = 0"
-            "AND agg_availing.registration_date::date < {start_month_string}"
-        ).format(start_month_string=start_month_string)
-        seeking_services = "NOT {not_registered} AND NOT {migrated}".format(
+        migrated = "(agg_migration.is_migrated = 1 AND agg_migration.migration_date::date < {start_month_string})::boolean".format(
+            start_month_string=start_month_string)
+        not_registered = "(agg_availing.is_registered = 0 AND agg_availing.registration_date::date < {start_month_string})::boolean".format(
+            start_month_string=start_month_string)
+        seeking_services = "NOT COALESCE({not_registered},'f') AND NOT COALESCE({migrated}, 'f')".format(
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py:99]: these coalesces are not necessary if you use `DISTINCT FROM` like you had before. why did you remove that?

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 8]
```diff
@@ -92,16 +92,16 @@ def _state_aggregation_query(self, state_id):
...
         ).format(start_month_string=start_month_string)
-        registered_status = (
-            "agg_availing.is_registered IS DISTINCT FROM 1"
+        not_registered = (
+            "agg_availing.is_registered = 0"
             "AND agg_availing.registration_date::date < {start_month_string}"
         ).format(start_month_string=start_month_string)
-        seeking_services = "NOT {registered_status} AND NOT {migration_status}".format(
-            registered_status=registered_status, migration_status=migration_status)
+        seeking_services = "NOT {not_registered} AND NOT {migrated}".format(
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: the double negative is a bit confusing "not not registered". It might be more straightforward to have the indicator be the opposite

**Review by calellowitz — CHANGES_REQUESTED** [review]
Few small things that i think will make this clearer. Plus still good to wrap those expressions in parentheses since it changes the logic to have those compound expressions dropped in without that

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py (hunk 7 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py:99]: you mentioned of using boolean instead of Integers that's why I am using this

**Comment by ShubhamBansal1997** [comment]
@calellowitz Shall I put only `seeking_services` in `parentheses` or the remaining ones (`migrated` and `not_registered`) also?

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 9]
```diff
@@ -96,7 +96,7 @@ def _state_aggregation_query(self, state_id):
             start_month_string=start_month_string)
         not_registered = "(agg_availing.is_registered = 0 AND agg_availing.registration_date::date < {start_month_string})::boolean".format(
             start_month_string=start_month_string)
-        seeking_services = "NOT COALESCE({not_registered},'f') AND NOT COALESCE({migrated}, 'f')".format(
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: sorry this wasnt clear. I meant using `distinct from` in the "not_registered" and "migrated" definitions. that way they can be dropped in here directly, and  will always be either true or false

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 10]
```diff
@@ -96,8 +96,10 @@ def _state_aggregation_query(self, state_id):
             start_month_string=start_month_string)
         not_registered = "(agg_availing.is_registered = 0 AND agg_availing.registration_date::date < {start_month_string})::boolean".format(
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py:97]: this should not need to be coerced to a boolean. those expressions already return a boolean

**Review by calellowitz — CHANGES_REQUESTED** [review]
Left a couple comments plus the double negative one from before is still true

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py (hunk 9 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: No, there are not always `true` or `false`, case of `NULL` is also possible in case the forms are not submitted, that's why using `DISTINCT`

**Comment by ShubhamBansal1997** [comment]
@calellowitz Updated the PR!!

**Review by calellowitz — APPROVED** [review]
Thanks for working through all that

**Review by Rohit25negi — APPROVED** [review]
Looks good.
